delivery/media/watermarking/lib/media-delivery-watermarking.js

```diff
@@ -0,0 +1,2134 @@
+                            if ("number" != typeof ranges && "bytes" === ranges.type && 1 === ranges.length) return ranges[0];
+                            throw new Error("Invalid range format!!");
+                        }
+                        throw new Error("Range header should be string!!");
+                    }(rangeHeader);
+                    position = CborParser.findPosition(sidecarObject, BigInt(range.start), BigInt(range.end));
+                } else position = CborParser.findPosition(sidecarObject, BigInt(-1), BigInt(-1), filename);
+                if (void 0 === position) throw new Error("Unable to find position from the side car file!");
+                logger.log("D:pos: %s", position);
+                let subPath, tmidVariant = 0;
+                if (-1 == position) subPath = this.getSubVariantPath(variantSubPath, tmidVariant); else {
+                    const cacheKey = await buildKey([ payload, secretKey ]);
+                    tmid = Cache.getTmid(cacheKey), tmid || (tmid = await IrdetoAlgorithm.generateTmid(payload.wmid, payload.wmopid, payload.wmidfmt, payload.wmpatlen, secretKey), 
+                    Cache.storeTmid(cacheKey, tmid));
+                    const tmidLenBits = 4 * tmid.length, tmidPos = position % tmidLenBits, tmidChar = tmid[tmidLenBits / 4 - Math.floor(tmidPos / 4) - 1], tmidBitPos = 1 << tmidPos % 4;
+                    tmidVariant = 0 == (parseInt(tmidChar, 16) & tmidBitPos) ? 0 : 1, subPath = this.getSubVariantPath(variantSubPath, tmidVariant);
+                }
+                if (null === subPath) throw new Error(`No watermarking subpath found for the variant ${tmidVariant}!`);
+                return `${basedir}/${subPath}/${filename}`;
+            }
+            throw new Error("Invalid wmvnd, indirect case with only 'irdeto' vendor is supported at the moment!");
+        }
+        throw new Error("Invalid wmidtyp, must be 0 (direct) or 1 (indirect)!");
+    }
+    async getSideCarObject(baseDir, filename) {
+        if (!baseDir || !filename) throw new Error("Unable to get side car object for the request, invalid url!");
+        const paceInfoResponse = await httpRequest(`${baseDir}/${Watermarking.WMPACEINFO_DIR}/${filename}`), contentLength = paceInfoResponse.getHeader("Content-Length");
+        if (null == contentLength || 0 === contentLength.length) throw new Error("Side car processing failed due to no content-length response header found!");
+        const paceinfoLength = parseInt(contentLength[0]), dataArr = await async function(stream, size) {
+            const buffer = new ArrayBuffer(size), arr = new Uint8Array(buffer);
+            let currentPos = 0;
+            return await stream.pipeTo(new WritableStream({
+                write(chunk) {
+                    arr.set(chunk, currentPos), currentPos += chunk.length;
+                }
+            })), arr;
+        }(paceInfoResponse.body, paceinfoLength);
+        return CborParser.decode(dataArr);
```

Comment:
Would it be safe/worthwhile to cache this? That would allow us to skip the fetch. 

I'm not familiar with the expected size of these files, so maybe it would require too much memory.

Reply:
@evan-hughes : Yes, even though cbor encoded files are much compact, the expected size for larger media assets would be around 55~60KB for one resolution. Incase we get request for 7 different resolution (adaptive streaming) we would end up with storage ~ 7 * ( 55KB). Also the side car file would have already cached as a part of sub-request by ghost i guess. So this storage would be unnecessary on EW right ?